syntax = "proto3";

package unary;
import "google/protobuf/struct.proto";

service workspace{
  // Workspace rpc declaration
 rpc UserSignup(UserData) returns (UserDataResponse) {};
 rpc UserLogin(UserLoginData) returns (UserLoginResponse) {};
 rpc CreateProject(ProjectData) returns (ProjectMinimalDetail) {};
 rpc CreateFolder(FolderData) returns (FolderMinimalDetail) {};
 rpc CreateFile(FileData) returns (FileMinimalDetail) {};
 rpc UpdateFile(FileData) returns (FileMinimalDetail) {};
 rpc ViewFile(FileData) returns (FileMinimalDetail) {};
 rpc ViewProject(ProjectData) returns (ProjectContent) {};

}

message UserData{
 string name = 1;
 string username = 2;
 string password = 3;
 string email = 4;

}

message UserDataResponse {
 string name = 1;
 string username = 2;
 string email = 3;
 string token = 4;
//  bool received = 5;
 int32 userId = 5;
 string message = 6;
 bool success = 7;
}

message ErrorResponse {
  string message = 1;
}

message UserLoginData {
  string username = 1;
  string password = 2;
}

message UserLoginResponse {
  string username = 1;
  string password = 2;
  string token = 3;
  string message = 4;
  int32 userId = 5;
  bool success = 6;
}

message ProjectData {
  string name = 1;
  string token = 2;
  int32 id = 3;
}

message ProjectMinimalDetail {
  string name = 1;
  int32 ownerId = 2;
  int32 id = 3;
  string message = 4;
  bool success = 5;
}

message ProjectContent {
  string name = 1;
  int32 id = 2;
  google.protobuf.Struct content = 3;
  string message = 4;
  bool success = 5;
}

message FolderData {
  string name = 1;
  // Either folder id or
  // project id
  int32 parentFolderId = 2;
  int32 projectId = 3;
  string token = 4;
}

message FolderMinimalDetail {
  string name = 1;
  // Either folder id or
  // project id
  int32 parentFolderId = 2;
  int32 projectId = 3;
  int32 ownerId = 4;
  int32 id = 5;
  string message = 6;
  bool success = 7;
}

message FileData {

  string name = 1;
  string token = 2;
  int32 folderId = 3;
  string content = 4;
  int32 id = 5;
}

message FileMinimalDetail {
  string name = 1;
  int32 ownerId = 2;
  int32 folderId = 3;
  int32 id = 4;
  string message = 5;
  bool success = 6;
  string content = 7;

}
